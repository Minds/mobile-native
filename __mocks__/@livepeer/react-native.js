// mock @livepeer/react-native
export const LivepeerConfig = {
  setConfig: jest.fn(),
  getConfig: jest.fn(),
  get: jest.fn(),
  set: jest.fn(),
};

export const createReactClient = jest.fn();
export const createLivepeerClient = jest.fn();
export const Player = jest.fn();
