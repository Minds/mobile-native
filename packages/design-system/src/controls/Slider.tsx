import { Slider as TSlider, SliderProps } from '@tamagui/slider';

export const Slider = (props: SliderProps) => (
  <TSlider size="$1.5" {...props}>
    <TSlider.Track backgroundColor="$grey-300">
      <TSlider.TrackActive backgroundColor="$action" />
    </TSlider.Track>
    <TSlider.Thumb
      circular
      index={0}
      borderColor="$colorTextPrimary"
      backgroundColor="$colorTextPrimary"
    />
  </TSlider>
);
