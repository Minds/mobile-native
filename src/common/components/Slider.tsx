import React from 'react';
import { StyleProp, View, ViewStyle } from 'react-native';
import { Slider as ValueSlider } from 'react-native-elements';
import MText from '~/common/components/MText';
import ThemedStyles from '~/styles/ThemedStyles';
import { B2, B3, Row } from '../ui';

const THUMB_TOUCH_SIZE = {
  width: 80,
  height: 80,
};

type PropsType = {
  title?: string;
  stepSize: number;
  defaultValue: number;
  currentValue: number;
  maximumRangeValue: number;
  minimumRangeValue: number;
  minimumStepLabel: string;
  maximumStepLabel: string;
  onAnswer: (number) => void;
  floatingLabel?: boolean;
  steps?: number[];
  formatValue?: (txt: number) => string;
};

const Slider = ({
  onAnswer,
  floatingLabel,
  steps,
  formatValue,
  ...props
}: PropsType) => {
  const value =
    typeof props.currentValue === 'number'
      ? props.currentValue
      : props.defaultValue;
  const completionPercentage =
    value === props.minimumRangeValue
      ? 0
      : (value / (props.maximumRangeValue || 100)) * 93;
  const labelSpacerStyle: StyleProp<ViewStyle> = {
    width: completionPercentage + '%',
    maxWidth: '90%',
    height: 20,
    flexDirection: 'row',
    position: 'relative',
  };
  const formattedValue = formatValue ? formatValue(value) : value;

  const handleValueChange = (newValue: number) => {
    if (steps) {
      onAnswer(findClosest(steps, newValue));
    } else {
      onAnswer(newValue);
    }
  };

  return (
    <View style={styles.container}>
      {!!props.title && <MText style={styles.title}>{props.title}</MText>}
      {floatingLabel && (
        <Row top="S">
          <View style={labelSpacerStyle} />
          <B3 color="secondary" align="center">
            {formattedValue}
          </B3>
        </Row>
      )}

      <ValueSlider
        value={value}
        onValueChange={handleValueChange}
        thumbTintColor={ThemedStyles.getColor('PrimaryText')}
        minimumValue={props.minimumRangeValue || 0}
        allowTouchTrack={false}
        step={props.stepSize}
        maximumValue={props.maximumRangeValue || 100}
        thumbTouchSize={THUMB_TOUCH_SIZE}
        maximumTrackTintColor={ThemedStyles.getColor('PrimaryBorder')}
        minimumTrackTintColor={ThemedStyles.getColor('Link')}
      />
      <View style={styles.textContainer} pointerEvents="none">
        <B2 color="secondary" font="regular">
          {props.minimumStepLabel}
        </B2>
        <B2 color="secondary" font="regular">
          {props.maximumStepLabel}
        </B2>
      </View>
    </View>
  );
};

const styles = ThemedStyles.create({
  container: ['marginBottom7x'],
  title: ['fontL', 'fontMedium'],
  textContainer: ['rowJustifySpaceBetween'],
  text: [{ fontSize: 13 }, 'fontMedium', 'colorSecondaryText'],
});

export default Slider;

function findClosest(arr: number[], num: number) {
  let minDiff = 1000;
  let ans;
  for (const i in arr) {
    var m = Math.abs(num - arr[i]);
    if (m < minDiff) {
      minDiff = m;
      ans = arr[i];
    }
  }
  return ans;
}
