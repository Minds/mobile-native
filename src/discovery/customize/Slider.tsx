import React from 'react';
import { View, Text } from 'react-native';
import i18n from '../../common/services/i18n.service';
import ThemedStyles from '../../styles/ThemedStyles';
import { Slider as ValueSlider } from 'react-native-elements';
import { useLocalStore } from 'mobx-react';

type data = {
  id: string;
  title: string;
  value: number;
  leftText: string;
  rightText: string;
};

type PropsType = {
  data: data;
};

const Slider = ({ data }: PropsType) => {
  const store = useLocalStore(() => ({
    question: data,
    onChange(value: number) {
      this.question.value = value;
    },
    get title() {
      return i18n.tf(
        `discovery.customizeQuestions.${this.question.id}.title`,
        this.question.title,
      );
    },
    get leftText() {
      return i18n.tf(
        `discovery.customizeQuestions.${this.question.id}.leftText`,
        this.question.leftText,
      );
    },
    get rightText() {
      return i18n.tf(
        `discovery.customizeQuestions.${this.question.id}.rightText`,
        this.question.rightText,
      );
    },
  }));
  return (
    <View style={styles.container}>
      <Text style={styles.title}>{store.title}</Text>
      <ValueSlider
        value={store.question.value}
        onSlidingComplete={store.onChange}
        thumbTintColor={ThemedStyles.getColor('Link')}
        minimumValue={0}
        maximumValue={100}
        maximumTrackTintColor={ThemedStyles.getColor('PrimaryBorder')}
        minimumTrackTintColor={ThemedStyles.getColor('PrimaryBorder')}
      />
      <View style={styles.textContainer}>
        <Text style={styles.text}>{store.leftText}</Text>
        <Text style={styles.text}>{store.rightText}</Text>
      </View>
    </View>
  );
};

const styles = ThemedStyles.create({
  container: ['marginBottom7x'],
  title: ['fontL', 'fontMedium'],
  textContainer: ['rowJustifySpaceBetween'],
  text: [{ fontSize: 13 }, 'fontMedium', 'colorSecondaryText'],
});

export default Slider;
