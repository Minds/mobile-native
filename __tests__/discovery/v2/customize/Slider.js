import React from 'react';
import 'react-native';
import { render, waitFor } from '@testing-library/react-native';
import Slider from '../../../../src/discovery/customize/Slider';

const mockData = {
  id: 'coso',
  title: 'Political Content',
  value: 50,
  leftText: 'LESS',
  rightText: 'MORE',
};

describe('Slider component', () => {
  it('should render correctly', () => {
    const { toJSON } = render(<Slider data={mockData} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
