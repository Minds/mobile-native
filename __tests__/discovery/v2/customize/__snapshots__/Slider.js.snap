// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider component should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "marginBottom": 35,
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 16,
        },
        Object {
          "fontFamily": "Roboto-Medium",
          "fontWeight": "500",
        },
      ]
    }
  >
    Political Content
  </Text>
  <View
    allowTouchTrack={false}
    animationType="timing"
    onLayout={[Function]}
    onSlidingComplete={[Function]}
    step={0}
    style={
      Object {
        "height": 40,
        "justifyContent": "center",
      }
    }
    theme={
      Object {
        "colors": Object {
          "black": "#242424",
          "disabled": "hsl(208, 8%, 90%)",
          "divider": "#bcbbc1",
          "error": "#ff190c",
          "grey0": "#393e42",
          "grey1": "#43484d",
          "grey2": "#5e6977",
          "grey3": "#86939e",
          "grey4": "#bdc6cf",
          "grey5": "#e1e8ee",
          "greyOutline": "#bbb",
          "platform": Object {
            "android": Object {
              "error": "#f44336",
              "grey": "rgba(0, 0, 0, 0.54)",
              "primary": "#2196f3",
              "secondary": "#9C27B0",
              "success": "#4caf50",
              "warning": "#ffeb3b",
            },
            "ios": Object {
              "error": "#ff3b30",
              "grey": "#7d7d7d",
              "primary": "#007aff",
              "searchBg": "#dcdce1",
              "secondary": "#5856d6",
              "success": "#4cd964",
              "warning": "#ffcc00",
            },
          },
          "primary": "#2089dc",
          "searchBg": "#303337",
          "secondary": "#ca71eb",
          "success": "#52c41a",
          "warning": "#faad14",
          "white": "#ffffff",
        },
      }
    }
    thumbTouchSize={
      Object {
        "height": 40,
        "width": 40,
      }
    }
    value={50}
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "backgroundColor": "#D3DBE3",
          "borderRadius": 2,
          "height": 4,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "#D3DBE3",
          "borderRadius": 2,
          "height": 0,
          "marginTop": -0,
          "position": "absolute",
          "width": 0,
        }
      }
    />
    <View
      onLayout={[Function]}
      style={
        Object {
          "backgroundColor": "#1B85D6",
          "borderRadius": 10,
          "height": 20,
          "position": "absolute",
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
          "width": 20,
        }
      }
      testID="sliderThumb"
    />
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "backgroundColor": "transparent",
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 13,
          },
          Object {
            "fontFamily": "Roboto-Medium",
            "fontWeight": "500",
          },
          Object {
            "color": "#72727C",
          },
        ]
      }
    >
      LESS
    </Text>
    <Text
      style={
        Array [
          Object {
            "fontSize": 13,
          },
          Object {
            "fontFamily": "Roboto-Medium",
            "fontWeight": "500",
          },
          Object {
            "color": "#72727C",
          },
        ]
      }
    >
      MORE
    </Text>
  </View>
</View>
`;
